Deploy SmartPointers in NativePromise.h

anil-hebbar opened the pull request:
#### ffa3482a154f6a6dda81d4c6bfcc2b9616dcdff4
<pre>
Deploy SmartPointers in NativePromise.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=287594">https://bugs.webkit.org/show_bug.cgi?id=287594</a>

Reviewed by Ryosuke Niwa.

Partly Addressed the smart pointer static analyzer warnings in NativePromise.h

* Source/WTF/wtf/NativePromise.h:
(WTF::NativePromiseRequest::disconnect):

Canonical link: <a href="https://commits.webkit.org/290491@main">https://commits.webkit.org/290491@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b0cfc31b97c229ae17a5b654960ac5e429fd4e2c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/90029 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9558 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44959 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/95029 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40802 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9945 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17882 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/69309 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26908 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/93030 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7607 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81695 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49671 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/7334 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/36054 "Found 8 new test failures: fast/html/process-end-tag-for-inbody-crash.html fonts/monospace.html fonts/sans-serif.html fonts/serif.html fullscreen/fullscreen-cancel-after-request-crash.html media/modern-media-controls/fullscreen-button/fullscreen-button.html media/modern-media-controls/invalid-placard/invalid-placard-constrained-metrics.html media/modern-media-controls/placard-support/placard-support-airplay-fullscreen.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39936 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/82832 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77668 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/37122 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96855 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/88807 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/17217 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12656 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/78303 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17473 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77519 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77509 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19173 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21964 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/20572 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/10411 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/17227 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/22552 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/111298 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16968 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26643 "Found 6 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.default, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager-jettison, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-slow-memory, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-collect-continuously, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-eager-jettison, wasm.yaml/wasm/stress/repro_1289.js.wasm-collect-continuously (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20420 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18751 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -320,10 +320,11 @@ class NativePromiseRequest :  public CanMakeWeakPtr<NativePromiseRequest> {
+        if (!safeCallback)
             return;
-        std::exchange(m_callback, nullptr)->disconnect();
+        std::exchange(safeCallback, nullptr)->disconnect();
```
This is unsafe and is likely why the tests are failing. The original code was setting the member variable `m_callback` to `nullptr`. But with this change, we're setting the temporary variable `safeCallback` to nullptr and `m_callback` remains as is. 

The warning https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/290324@main%20(1936)/scan-build-output/StaticAnalyzer/WTF/StaticAnalyzerReports/report-NativePromise.h-disconnect-4-a52b14.html#EndPath is not complaining that the use if `m_callback` is unsafe. It's complaining that the use of `std::exchange(m_callback, nullptr)` is unsafe. 

What you want here is: 

`void disconnect()`
`    {`
`      ASSERT(m_callback);`
`       if (!m_callback)`
`           return;`
`       RefPtr { std::exchange(m_callback, nullptr) }->disconnect();`
`   }`

RupinMittal commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -940,7 +941,7 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
             ASSERT(m_thenCallback, "Can only track a request once");
             requestHolder.track(*m_thenCallback);
-            m_promise->maybeSettle(m_thenCallback.releaseNonNull(), m_logSiteIdentifier);
+            protectedPromise()->maybeSettle(m_thenCallback.releaseNonNull(), m_logSiteIdentifier);
```
There are no static analyzer warnings about `m_promise` or `m_thenCallback` in NativePromise.h. Could you explain why we're making these changes?

RupinMittal replied:
Ah it seems like this is a templated class and so we need to look at the failures in WebKitBuild/. So we do want these changes.

RupinMittal reviewed (commented):
The only warning I see in NativePromise.h is https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/290324@main%20(1936)/scan-build-output/StaticAnalyzer/WTF/StaticAnalyzerReports/report-NativePromise.h-disconnect-4-a52b14.html#EndPath. What's the reason for the other changes?

RupinMittal commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -955,17 +956,28 @@ class NativePromise final : public NativePromiseBase, public ConvertibleToNative
+            return m_thenCallback;
+        }
+
+        Ref<NativePromise> protectedPromise()
```
`m_promise` is never re-assigned, so we should mark the member variable `const` instead of making a `protectedPromise` function.

RupinMittal commented on Source/WTF/wtf/NativePromise.h:
```diff
@@ -1513,19 +1525,23 @@ class NativePromiseProducer final : public ConvertibleToNativePromise {
     // The Producer may be moved to resolve/reject the completion promise.
     // While we expect m_promise to never be null, it would cause a null dereference in the destructor if the destructor was called after a move.
     RefPtr<PromiseType> m_promise;
+    RefPtr<PromiseType> protectedPromise() const
```
m_promise is never re-assigned, so we should mark the member variable const instead of making a protectedPromise function.

anil-hebbar committed:
Deploy SmartPointers in NativePromise.h
https://bugs.webkit.org/show_bug.cgi?id=287594

Reviewed by Ryosuke Niwa.

Partly Addressed the smart pointer static analyzer warnings in NativePromise.h

* Source/WTF/wtf/NativePromise.h:
(WTF::NativePromiseRequest::disconnect):

Canonical link: https://commits.webkit.org/290491@main
